src/wallet/wallet.cpp

```diff
@@ -2008,9 +2053,8 @@ bool CWallet::SelectorMatchesAddress(
             return a0 == a1;
         },
         [&](const libzcash::SaplingExtendedFullViewingKey& extfvk) {
-            auto j = extfvk.DecryptDiversifier(a0.d);
-            auto addr = extfvk.Address(j);
-            return addr.has_value() && addr.value() == a0;
+            auto j = extfvk.DecryptDiversifier(a0);
```

Comment:
Diversifier encryption is not authenticated, so won't it just produce the wrong `j` when decrypting with the wrong key?

Reply:
If you look at the implementation of `DecryptDiversifier`, it now attempts to reproduce the original address, and returns std::nullopt if it does not succeed.

Reply:
Oh, `DecryptDiversifier` is doing more than I thought. Maybe rename it to `DecryptAndCheckDiversifier`?